fdk.go

```diff
@@ -107,6 +111,84 @@ func doHTTP(handler Handler, ctx context.Context, in io.Reader, out io.Writer) {
 	}
 }
 
+func doJSON(handler Handler, ctx context.Context, in io.Reader, out io.Writer) {
+	var buf bytes.Buffer
+	hdr := make(http.Header)
+
+	for {
+		doJSONOnce(handler, ctx, in, out, &buf, hdr)
+	}
+}
+
+type jsonio struct {
+	Body        string `json:"body"`
+	ContentType string `json:"content_type"`
+}
+
+type callRequestHTTP struct {
+	Type       string      `json:"type"`
+	RequestURL string      `json:"request_url"`
+	Headers    http.Header `json:"headers"`
+}
+
+type jsonIn struct {
+	jsonio
+	CallID   string           `json:"call_id"`
+	Protocol *callRequestHTTP `json:"protocol"`
+}
+
+type callResponseHTTP struct {
+	StatusCode int         `json:"status_code,omitempty"`
+	Headers    http.Header `json:"headers,omitempty"`
+}
+
+type jsonOut struct {
+	jsonio
+	Protocol *callResponseHTTP `json:"protocol,omitempty"`
+
+	io.Writer
+}
+
+func (out *jsonOut) Write(p []byte) (n int, err error) {
```

Comment:
the methods around this are a little confusing. one nit is that this [and jsonIn] would be slightly more readable if all the fields were all here instead of inheriting a mix of embedded fields from `jsonio` as well as any methods on `io.Writer`.

this writer seems unused since this struct overwrites `io.Writer` method and does not use the `io.Writer` inside of it. this writer (buffer) seems best used in the caller and not embedded in this struct, to accumulate a body from the user into it (in the caller), which can then be inserted in the `*jsonOut` body as a string once (in the caller), instead of doing many string concatenations (which each generate a new string, strings are immutable). i realize this was handling status code setting as well. personally i like having models that are just models, but maybe that is just me; it would be possible to have another simply type to allow setting status code and writing to the buffer, which would then be translated to a json encodable struct, but having these in the same struct is not the end of the world. at present, however, we need to stop overwriting the Write method and bypassing the buffer, and set the body in doJSONOnce before encoding this struct.